### PR TITLE
Homogenized Universe

### LINKED ISSUE #11: Support: Homogenized Universe Object
Implement a supporting object that will be used by the branching reader and results reader that stores data for a single homogenized universe at a single instance in time.

## Requirements
1. Should store the universe number, and, optionally, day, burnup, and burnup step for this instance. Optional because for a problem with no burnup, those values will not be present in the file.
1. Simple methods to add and get variable data, that will be sent from the readers.

## Optional
1. Output the data with some template engine like [Jinja2](http://jinja.pocoo.org/) or [Mako](http://www.makotemplates.org/)

## Holding up
#6
#7

### PR DESCRIPTION
**Pull request by sallustius**
Container for storing and getting data.
Closes #11

### CODE REVIEW

File: serpentTools/objects/containers.py [hunk 1]
```diff
@@ -0,0 +1,92 @@
...
+        setter = self._lookup(variablename, uncertainty)
+        # 3. Return the value of the variable
+        x = setter.get(variablename)
+        if not uncertainty:
+            return x
+        else:
+            dx = setter.get(variablename)
+            return x, dx
+
+    def _lookup(self,variablename, uncertainty):
```
Comment (drewejohnson) [inline serpentTools/objects/containers.py]: Single space after items in lists/tuples/functions/etc e.g.
`_lookup(self, variableName, uncertainty):`
Mixed case variables as well

File: serpentTools/objects/containers.py [hunk 2]
```diff
@@ -0,0 +1,92 @@
...
+                return self.infExp
+            else:
+                return  self.infUnc
+        elif "b1" in variablename:
+            if not uncertainty:
+                return self.b1Exp
+            else:
+                return self.b1Unc
+
+        messages.error('Neither inf, nor b1 in the string')
```
Comment (drewejohnson) [inline serpentTools/objects/containers.py]: Two items here.

- Like the above error with the bad `uncertainty` type, the name of the variable should be passed into this error. This way, the user knows what our program is interpreting.
- Last line in the file should be blank

File: serpentTools/objects/containers.py [hunk 3]
```diff
@@ -0,0 +1,92 @@
+import numpy
```
Comment (drewejohnson) [inline serpentTools/objects/containers.py]: The `numpy` module isn't used here, so this line can be removed.

Further, could you add a brief [module level docstring](https://github.com/numpy/numpy/blob/master/doc/HOWTO_DOCUMENT.rst.txt#documenting-modules) at the top of this file? It largely just needs to indicate the purpose of the file and classes present (to be expanded as we go)

File: serpentTools/objects/containers.py (hunk 1 above)
Comment (sallustius) [reply serpentTools/objects/containers.py]: Done

File: serpentTools/objects/containers.py [hunk 4]
```diff
@@ -0,0 +1,92 @@
...
+        if not isinstance(uncertainty, bool):
+            raise messages.error("Uncertainty must be a boolean variable")
+        # 2. Pointer to the proper dictionary
+        setter = self._lookup(variablename, uncertainty)
+        # 3. Return the value of the variable
+        x = setter.get(variablename)
+        if not uncertainty:
+            return x
+        else:
+            dx = setter.get(variablename)
```
[the remark this answers is not in the document]
Comment (sallustius) [reply serpentTools/objects/containers.py]: Done

File: serpentTools/objects/containers.py [hunk 5]
```diff
@@ -0,0 +1,92 @@
...
+    ----------
+    -name: name of the universe
+    -bu:   burnup value
+    -step: temporal step
+    -day:  depletion day
+
+    """
+
+    def __init__(self, container, name, bu, step, day):
+        """ Class Initializer. Each universe is defined  uniquely
```
[the remark this answers is not in the document]
Comment (sallustius) [reply serpentTools/objects/containers.py]: Done

File: serpentTools/objects/containers.py [hunk 6]
```diff
@@ -0,0 +1,92 @@
...
+        self.step = step
+        self.day = day
+        # Dictionaries:
+        self.b1Exp = {}
+        self.infExp = {}
+        self.b1Unc = {}
+        self.infUnc = {}
+
+    def set(self, variablename, variablevalue, uncertainty=False):
+        """
```
[the remark this answers is not in the document]
Comment (sallustius) [reply serpentTools/objects/containers.py]: Done

File: serpentTools/objects/containers.py [hunk 7]
```diff
@@ -0,0 +1,92 @@
+import numpy
+from serpentTools.objects import _SupportingObject
+from serpentTools.settings import messages
+
+
+class HomogUniv(_SupportingObject):
+    """Class for:
```
[the remark this answers is not in the document]
Comment (sallustius) [reply serpentTools/objects/containers.py]: Done the first point

File: serpentTools/objects/containers.py [hunk 8]
```diff
@@ -0,0 +1,92 @@
...
+            raise messages.error("Uncertainty must be a boolean variable")
+        # 2. Pointer to the proper dictionary
+        setter = self._lookup(variablename, uncertainty)
+        # 3. Check if variable is already present. Then set the variable.
+        if variablename in setter:
+            messages.warning('The variable will be overwritten')
+        setter[variablename] = variablevalue
+
+    def get(self, variablename, uncertainty=False):
+
```
[the remark this answers is not in the document]
Comment (sallustius) [reply serpentTools/objects/containers.py]: Done

File: serpentTools/objects/containers.py [hunk 9]
```diff
@@ -0,0 +1,92 @@
...
+        # 1. Check the input values
+        variablename = _SupportingObject._convertVariableName(variablename)
+        if not isinstance(uncertainty, bool):
+            raise messages.error("Uncertainty must be a boolean variable")
+        # 2. Pointer to the proper dictionary
+        setter = self._lookup(variablename, uncertainty)
+        # 3. Return the value of the variable
+        x = setter.get(variablename)
```
[the remark this answers is not in the document]
Comment (sallustius) [reply serpentTools/objects/containers.py]: Done

File: serpentTools/objects/containers.py (hunk 3 above)
Comment (sallustius) [reply serpentTools/objects/containers.py]: Done

File: serpentTools/objects/containers.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (drewejohnson) [reply serpentTools/objects/containers.py]: The `__str__` issue will be addressed in the process of closing #36

File: serpentTools/objects/containers.py [hunk 10]
```diff
@@ -0,0 +1,92 @@
...
+        variablename:   Variable Name
+        variablevalue:  Variable Expected Value
+        uncertainty:    Boolean Variable- set to True in order to retrieve the
+                        uncertainty associated to the expected values
+        """
+
+        # 1. Check the input type
+        variablename = _SupportingObject._convertVariableName(variablename)
+        if not isinstance(uncertainty, bool):
+            raise messages.error("Uncertainty must be a boolean variable")
```
[the remark this answers is not in the document]
Comment (sallustius) [reply serpentTools/objects/containers.py]: Done

File: serpentTools/objects/containers.py [hunk 11]
```diff
@@ -0,0 +1,92 @@
...
+        if variablename in setter:
+            messages.warning('The variable will be overwritten')
+        setter[variablename] = variablevalue
+
+    def get(self, variablename, uncertainty=False):
+
+        # 1. Check the input values
+        variablename = _SupportingObject._convertVariableName(variablename)
+        if not isinstance(uncertainty, bool):
+            raise messages.error("Uncertainty must be a boolean variable")
```
[the remark this answers is not in the document]
Comment (sallustius) [reply serpentTools/objects/containers.py]: Done

File: serpentTools/objects/containers.py (hunk 2 above)
Comment (sallustius) [reply serpentTools/objects/containers.py]: Done

File: serpentTools/objects/containers.py [hunk 12]
```diff
@@ -39,54 +50,84 @@ def __init__(self, container, name, bu, step, day):
...
-            raise messages.error("Uncertainty must be a boolean variable")
+            raise messages.error('The variable uncertainty has type %s.\n ...'
+                                 'It should be boolean.', type(uncertainty))
         # 2. Pointer to the proper dictionary
-        setter = self._lookup(variablename, uncertainty)
+        setter = self._lookup(variableName, uncertainty)
         # 3. Check if variable is already present. Then set the variable.
-        if variablename in setter:
+        if variableName in setter:
             messages.warning('The variable will be overwritten')
```
Comment (drewejohnson) [inline serpentTools/objects/containers.py]: Add the name of the variable to this line

File: serpentTools/objects/containers.py [hunk 13]
```diff
@@ -39,54 +50,84 @@ def __init__(self, container, name, bu, step, day):
...
+        -------
+        x: str/tuple/list
+            Variable Value
+        dx: str/tuple/list
+            Associated uncertainty
+
+        """
         # 1. Check the input values
-        variablename = _SupportingObject._convertVariableName(variablename)
+        variableName = _SupportingObject._convertVariableName(variableName)
```
Comment (drewejohnson) [inline serpentTools/objects/containers.py]: Please remove.
This is only necessary if the variable name requested is SERPENT_STYLE. We will make it known to the user that the variables are __stored__ with mixedCase names.

File: serpentTools/objects/containers.py [hunk 14]
```diff
@@ -39,54 +50,84 @@ def __init__(self, container, name, bu, step, day):
...
+                return self.infUnc
+        elif "b1" in variableName:
             if not uncertainty:
                 return self.b1Exp
             else:
                 return self.b1Unc
 
-        messages.error('Neither inf, nor b1 in the string')
+        messages.error('%s not found.The string must contain either the   ...'
+                       'inf, or b1 substring.', variableName)
```
Comment (drewejohnson) [inline serpentTools/objects/containers.py]: Option to add other data like group structure to the object?

**Comment by gridley** [comment]
Python doesn't have any private data, so why are you guys using setters and getters?

File: serpentTools/objects/containers.py [hunk 15]
```diff
@@ -0,0 +1,144 @@
...
+        -------
+        x: str/tuple/list
+            Variable Value
+        dx: str/tuple/list
+            Associated uncertainty
+
+        """
+        # 1. Check the input values
+        if not isinstance(uncertainty, bool):
+            raise messages.error('The variable uncertainty has type %s.\n ...'
```
Comment (drewejohnson) [inline serpentTools/objects/containers.py]: Have this raise a [`TypeError`](https://docs.python.org/2/library/exceptions.html#exceptions.TypeError) with the same message instead. Same with the call in the `addData` method

File: serpentTools/tests/test_branching.py [hunk 16]
```diff
@@ -11,7 +11,7 @@
 
 
 class BranchTester(unittest.TestCase):
-    """Classs for testing the branching coefficient file reader."""
+    """Class for testing the branching coefficient file reader."""
```
Comment (drewejohnson) [inline serpentTools/tests/test_branching.py]: 👍

File: serpentTools/tests/test_container.py [hunk 17]
```diff
@@ -0,0 +1,86 @@
+"""Test the container object. """
+
+import unittest
+from serpentTools.objects import containers
+from serpentTools.parsers import DepletionReader
+
+
+class containerTester(unittest.TestCase):
```
Comment (drewejohnson) [inline serpentTools/tests/test_container.py]: Change the name of this class to reflect the container being tested. We have/will have many containers down the road

File: serpentTools/tests/test_container.py [hunk 18]
```diff
@@ -0,0 +1,86 @@
...
+    """ Class to test the container object """
+
+    @classmethod
+    def setUpClass(cls):
+        container = DepletionReader(None)
```
Comment (drewejohnson) [inline serpentTools/tests/test_container.py]: Pedantic/non-crucial request: Can you just pass `DepletionReader(None)` as the first argument to `containers.HomogUniv`? There isn't a need to store the Depletion reader after the creation of the `HomogUniv`

File: serpentTools/tests/test_container.py [hunk 19]
```diff
@@ -0,0 +1,86 @@
...
+                      ([6, 7, 8], [.6, .7, .8])}
+        cls.meta = {'macroE': (.1e-1, .2e-1, .3e-1, .1e-1, .2e-1, .3e-1),
+                    'test1': 'addio'}
+        # Use addData
+        for kk in cls.Exp:
+            cls.univ.addData(kk, cls.Exp[kk], False)
+        for kk in cls.Unc:
+            cls.univ.addData(kk, cls.Unc[kk], True)
+
+    def test_1(self):
```
Comment (drewejohnson) [inline serpentTools/tests/test_container.py]: Rename the test to indicate the purpose of this test, i.e. `test_getB1Exp`

File: serpentTools/tests/test_container.py [hunk 20]
```diff
@@ -0,0 +1,86 @@
...
+        """ Test the get function on b1Exp"""
+        d = {}
+        # Compare b1 dictionary
+        for kk in self.univ.b1Exp:
+            d[kk] = self.univ.get(kk,False)
+        self.assertDictEqual(self.b1Exp,d)
+
+    def test_2(self):
```
Comment (drewejohnson) [inline serpentTools/tests/test_container.py]: Same as ⬆️

File: serpentTools/tests/test_container.py [hunk 21]
```diff
@@ -0,0 +1,86 @@
...
+        """ Test the get function on b1Exp and b1Unc"""
+        d = {}
+        # Compare b1 dictionary with uncertainty
+        for kk in self.univ.b1Exp:
+            d[kk] = self.univ.get(kk,True)
+        self.assertDictEqual(self.b1Unc,d)
+
+    def test_3(self):
```
Comment (drewejohnson) [inline serpentTools/tests/test_container.py]: Same as ⬆️

File: serpentTools/tests/test_container.py [hunk 22]
```diff
@@ -0,0 +1,86 @@
...
+        """ Test the get function on infExp"""
+        d = {}
+        # Compare Inf dictionary
+        for kk in self.univ.infExp:
+            d[kk] = self.univ.get(kk, False)
+        self.assertDictEqual(self.infExp, d)
+
+    def test_4(self):
```
Comment (drewejohnson) [inline serpentTools/tests/test_container.py]: Same as ⬆️

File: serpentTools/tests/test_container.py [hunk 23]
```diff
@@ -0,0 +1,86 @@
...
+        """ Test the get function on infExp and infUnc"""
+        d = {}
+        # Compare inf dictionary with uncertainty
+        for kk in self.univ.infUnc:
+            d[kk] = self.univ.get(kk, True)
+        self.assertDictEqual(self.infUnc, d)
+
+    def test_5(self):
```
Comment (drewejohnson) [inline serpentTools/tests/test_container.py]: Same as ⬆️

File: serpentTools/objects/containers.py (hunk 15 above)
Comment (sallustius) [reply serpentTools/objects/containers.py]: Done

File: serpentTools/tests/test_container.py (hunk 17 above)
Comment (sallustius) [reply serpentTools/tests/test_container.py]: Done. Changed in "HomogenizedUniverseTester"

File: serpentTools/tests/test_container.py (hunk 19 above)
Comment (sallustius) [reply serpentTools/tests/test_container.py]: Done

File: serpentTools/tests/test_container.py (hunk 20 above)
Comment (sallustius) [reply serpentTools/tests/test_container.py]: Done

File: serpentTools/tests/test_container.py (hunk 21 above)
Comment (sallustius) [reply serpentTools/tests/test_container.py]: Done

File: serpentTools/tests/test_container.py (hunk 22 above)
Comment (sallustius) [reply serpentTools/tests/test_container.py]: Done

File: serpentTools/tests/test_container.py (hunk 23 above)
Comment (sallustius) [reply serpentTools/tests/test_container.py]: Done

File: serpentTools/tests/test_container.py (hunk 18 above)
Comment (sallustius) [reply serpentTools/tests/test_container.py]: Done

**Pull request merged**